python/pyspark/sql/catalog.py

```diff
@@ -0,0 +1,426 @@
+        :param f: python function
+        :param returnType: a :class:`DataType` object
+
+        >>> spark.catalog.registerFunction("stringLengthString", lambda x: len(x))
+        >>> spark.sql("SELECT stringLengthString('test')").collect()
+        [Row(stringLengthString(test)=u'4')]
+
+        >>> from pyspark.sql.types import IntegerType
+        >>> spark.catalog.registerFunction("stringLengthInt", lambda x: len(x), IntegerType())
+        >>> spark.sql("SELECT stringLengthInt('test')").collect()
+        [Row(stringLengthInt(test)=4)]
+
+        >>> from pyspark.sql.types import IntegerType
+        >>> spark.udf.register("stringLengthInt", lambda x: len(x), IntegerType())
+        >>> spark.sql("SELECT stringLengthInt('test')").collect()
+        [Row(stringLengthInt(test)=4)]
+        """
+        udf = UserDefinedFunction(f, returnType, name)
+        self._jsparkSession.udf().registerPython(name, udf._judf)
+
+    @since(2.0)
+    def isCached(self, tableName):
+        """Returns true if the table is currently cached in-memory.
+
+        >>> spark.catalog._reset()
+        >>> spark.createDataFrame([(2, 2), (3, 3)]).registerTempTable("my_tab")
+        >>> spark.catalog.isCached("my_tab")
+        False
+        >>> spark.catalog.cacheTable("does_not_exist") # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+            ...
+        AnalysisException: ...
+        """
+        return self._jcatalog.isCached(tableName)
+
+    @since(2.0)
+    def cacheTable(self, tableName):
+        """Caches the specified table in-memory.
+
+        >>> spark.catalog._reset()
+        >>> spark.createDataFrame([(2, 2), (3, 3)]).registerTempTable("my_tab")
+        >>> spark.catalog.isCached("my_tab")
+        False
+        >>> spark.catalog.cacheTable("my_tab")
+        >>> spark.catalog.isCached("my_tab")
+        True
+        >>> spark.catalog.cacheTable("does_not_exist") # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+            ...
+        AnalysisException: ...
+        """
+        self._jcatalog.cacheTable(tableName)
+
+    @since(2.0)
+    def uncacheTable(self, tableName):
+        """Removes the specified table from the in-memory cache.
+
+        >>> spark.catalog._reset()
+        >>> spark.createDataFrame([(2, 2), (3, 3)]).registerTempTable("my_tab")
+        >>> spark.catalog.cacheTable("my_tab")
+        >>> spark.catalog.isCached("my_tab")
+        True
+        >>> spark.catalog.uncacheTable("my_tab")
+        >>> spark.catalog.isCached("my_tab")
+        False
+        >>> spark.catalog.uncacheTable("does_not_exist") # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+            ...
+        AnalysisException: ...
+        """
+        self._jcatalog.uncacheTable(tableName)
+
+    @since(2.0)
+    def clearCache(self):
+        """Removes all cached tables from the in-memory cache.
+
+        >>> spark.catalog._reset()
```

Comment:
These will be part of the API, so it's better to only have a few here as examples, others should be in unit tests.